host_core/native/hostcore_wasmcloud_native/src/wasmruntime.rs

```diff
@@ -0,0 +1,239 @@
+            token: CallbackToken {
+                continue_signal: Condvar::new(),
+                return_value: Mutex::new(None),
+            },
+        });
+
+        // TODO: figure out if we need the caller tokens
+        //let caller_token = set_caller(...);
+        //let caller_token = 12;
+
+        let mut msg_env = OwnedEnv::new();
+        msg_env.send_and_clear(&self.pid.clone(), |env| {
+            (
+                atoms::invoke_callback(),
+                crate::Claims::from(claims.clone()),
+                binding,
+                invocation.namespace,
+                invocation.operation,
+                invocation.payload.unwrap_or_default(),
+                callback_token.clone(),
+            )
+                .encode(env)
+        });
+
+        let mut result = callback_token.token.return_value.lock().unwrap();
+        while result.is_none() {
+            result = callback_token.token.continue_signal.wait(result).unwrap();
+        }
+
+        let result: &(bool, Vec<u8>) = result
+            .as_ref()
+            .expect("expect callback token to contain a result");
+        match result {
+            (true, return_value) => Ok(Some(return_value.clone())),
+            (false, _) => {
+                // TODO: verify whether we should return none here or use an Err
+                error!("Elixir callback threw an exception.");
+                Ok(None)
+            }
+        }
+    }
+}
+
+pub fn on_load(env: Env) -> bool {
+    rustler::resource!(RuntimeResource, env);
+    rustler::resource!(ActorResource, env);
+    true
+}
+
+#[rustler::nif(name = "runtime_new")]
+pub fn new<'a>(
+    env: rustler::Env<'a>,
+    _config: ExRuntimeConfig,
+) -> Result<ResourceArc<RuntimeResource>, rustler::Error> {
+    let host_handler = ElixirHandler { pid: env.pid() };
+    let rt = WcRuntime::from_host_handler(host_handler)
+        .context("failed to construct runtime")
+        .map_err(|e| Error::Term(Box::new(e.to_string())))?;
+
+    let resource = ResourceArc::new(RuntimeResource { inner: rt });
+    Ok(resource)
+}
+
+#[rustler::nif(name = "version")]
+pub fn version(runtime_resource: ResourceArc<RuntimeResource>) -> Result<String, rustler::Error> {
+    let v = runtime_resource.inner.version();
+
+    Ok(v.to_string())
+}
+
+/// Called from the Elixir native wrapper which is in turn wrapped by the Wasmcloud.Runtime.Server GenServer
+#[rustler::nif(name = "start_actor")]
+pub fn start_actor<'a>(
+    env: rustler::Env<'a>,
+    runtime_resource: ResourceArc<RuntimeResource>,
+    bytes: Binary<'a>,
+) -> Result<ResourceArc<ActorResource>, rustler::Error> {
+    let actor = Actor::new(&runtime_resource.inner, bytes.as_slice())
+        .context("failed to load actor from bytes")
+        .unwrap();
+
+    let ar = ActorResource { actor };
+
+    Ok(ResourceArc::new(ar))
+}
+
+// NOTE: wasmex uses dirtycpu here. should we use dirty IO ?
+#[rustler::nif(name = "call_actor", schedule = "DirtyCpu")]
```

Comment:
I actually don't think we need the DirtyCpu or DirtyIo here, spawning a thread likely always comes in under a millisecond and all execution is in that thread, which is separate from the erlang scheduler

Reply:
https://github.com/tessi/wasmex/blob/6594e2c1d592c0c58f72200d4eb2285b33c0f327/native/wasmex/src/instance.rs#L102-L103

Reply:
Never mind then 😄 I do think in principle DirtyIo makes sense here, I'd be interested in experimenting with that value since our Wasms usually take longer waiting for responses for NATS messages then they do computing things

Reply:
yeah it's definitely something to spike/benchmark on